cordapp-example/contracts-kotlin/src/main/kotlin/com/example/contract/IOUContract.kt

```diff
@@ -30,17 +30,32 @@ class IOUContract : Contract {
      * considered valid.
      */
     override fun verify(tx: LedgerTransaction) {
-        val command = tx.commands.requireSingleCommand<Commands.Create>()
-        requireThat {
-            // Generic constraints around the IOU transaction.
-            "No inputs should be consumed when issuing an IOU." using (tx.inputs.isEmpty())
-            "Only one output state should be created." using (tx.outputs.size == 1)
-            val out = tx.outputsOfType<IOUState>().single()
-            "The lender and the borrower cannot be the same entity." using (out.lender != out.borrower)
-            "All of the participants must be signers." using (command.signers.containsAll(out.participants.map { it.owningKey }))
-
-            // IOU-specific constraints.
-            "The IOU's value must be non-negative." using (out.value > 0)
+        val command = tx.commands.requireSingleCommand<Commands>()
+
+        when (command.value) {
+            is Commands.Create -> requireThat {
+                // Generic constraints around the IOU transaction.
+                "No inputs should be consumed when issuing an IOU." using (tx.inputs.isEmpty())
+                "Only one output state should be created." using (tx.outputs.size == 1)
+                val out = tx.outputsOfType<IOUState>().single()
+                "The lender and the borrower cannot be the same entity." using (out.lender != out.borrower)
+                "All of the participants must be signers." using (command.signers.containsAll(out.participants.map { it.owningKey }))
+
+                // IOU-specific constraints.
+                "The IOU's value must be non-negative." using (out.value > 0)
+            }
+            is Commands.Destroy -> requireThat {
+                "There should be one IOU input." using (tx.inputs.size == 1)
+                "There should be no output." using (tx.outputs.isEmpty())
+
+                val inputIOU = tx.inputsOfType<IOUState>().single()
+
+                "Expect two signers." using (command.signers.size == 2)
+                "All participants must be signers." using (command.signers.containsAll(inputIOU.participants.map { it.owningKey }))
```

Comment:
On destruction, the borrower loses a liability. Wouldn't the borrower be happy to lose this liability in any case? So is their signature really needed?
If the borrower needs to atomically pay with tokens too, then sure the borrower's signature is necessary. But here, mmh. What was your rationale?
The same could be said of the issuer's signature on `Create`.

Reply:
Now I'm starting seeing how corda works, but few days before with this first task I was confusing and my way of thinking was like this:
1. Lender start the flow and lend money to Borrower - I need both signatures because Lender must agree to lend and Borrower must confirm he accepted it
1. Now, when borrower want to payback, he must send "it" (so his signature required) and as a proof he would demand lender signature as confirmation. 

Now you suggesting, that Lender signature is enough... and I can see why, but let imagine this case: Borrower send tokens  to Lender and now his waiting for Lender to start a Payback flow and destroy IUO. What if Lender don't do it? 
I can imagine opposite - only Borrower sign payback (because he don't need permission from Lender) but he actually don't do actual payback. 

So I used two signatures because of above cases.

Reply:
You are grappling with the problems that come with 2 transactions. What you want to achieve is an atomic transaction. With an atomic operation, the borrower did not "send cash but the IOU is still there".

Reply:
Ok, then please explain me yet the following, because this is where I can't put things together:
1. I don't understand yet how to make "send cash/tokens" and "consume IOU" as atomic transaction in Corda. Should I treat `value` as token/coin/cash etc. here? Might be stupid questions but well... ;)  
1. if this is atomic, then shouldn't Borrower be the one who sign payback? After all it's Borrower who knows if and when he can do payback. If we require Lender siganture, then Lender need to "monitor" situation with Borrower. Or maybe (it just came to me) you ment, that Borrower start the flow, but Lender is the one who needs to sign it?

Reply:
1. To create an atomic transaction with IOU destroy and cash transfer from borrower you need to create a flow that will do that. This flow will:
    1. add the IOU in input
    2. add the IOU destroy command with lender as signer
    3. add a cash input owned by the borrower
    4. add a cash output owned by the lender
    5. add the cash transfer command with borrower as signer
    6. verify, sign, finalise
    7. Of course lender and borrower won't accept to sign if the cash amount does not match the IOU amount.

    So what we have then is a tx with 2 commands, 2 inputs and 1 output. Does your 2 contracts allow that?

    * If your IOU Destroy expects `.inputs.size == 1` then no
    * If your IOU Destroy expects `.inputsOfType<IOUState>.size == 1` then yes
    * If your IOU Destroy expects `.outputs.size == 0` then no.
    * If your IOU Destroy expects `.outsputsOfType<IOUState>.size == 0` then yes
2. Do not conflate the 2 commands. See my points 2 and 5.

    * Borrower loses the asset cash, Lender loses the asset IOU.
    * So they need to both sign on the transaction as a whole, but they are **required** to sign by the respective commands where they lose an asset.
    * It does not matter who starts the flow here. I would say it is better to start with the borrower because they both know the IOU state but only the borrower known which cash state will be used.

Reply:
Thanks for detailed explanation, but I have yet two questions:
1. `cash transfer command` - in current implementation I don't have any cash transfer command, so I should add new type of command ie `class Transfer : Commands`, and then add it to the transaction `.addCommand(Command(IOUContract.Commands.Transfer(), iouState.borrower.owningKey))`
1. `Does your 2 contracts allow that?` - there is only one contract... do you meant 2 commands?

Basically what I still struggle to understand is: where is **a money** here? We have `Create` that creates IOU with some value. But this value is not money, right? Or is it?   
Why in `ExampleFlow` we don't have any transfer from lender -> borrower but we need some transfer for `PaybackFlow`?

I feel like I either I do not understand what the task is really about or I totally not get yet how Corda works... or probably both.

Reply:
1. When you have a cash transfer, you will do: `.addCommand(Command(CashContract.Commands.Transfer(), cashState.holder.owningKey))` <- only mention of cash, not of IOU.
2. There will be a cash contract too. Your cash state will be protected by the cash contract. So the atomic tx will have 2 contracts.

The IOU is not money. It is an asset. We don't have a money transfer in ExampleFlow because it was made very simple but in reality, it is desirable to have that happen at the same time. 
You are conflating what is ultimately desirable and what has been coded so far.

Reply:
I updated a code - I hope I'm getting in a right direction.